pkg/apis/trainer/v1alpha1/trainjob_types.go

```diff
@@ -232,12 +232,37 @@ type Trainer struct {
 	NumProcPerNode *intstr.IntOrString `json:"numProcPerNode,omitempty"`
 }
 
-// PodSpecOverride represents the custom overrides that will be applied for the TrainJob's resources.
-type PodSpecOverride struct {
+// PodTemplateOverride represents the custom overrides that will be applied for the TrainJob's resources.
+type PodTemplateOverride struct {
 	// TrainJobs is the training job replicas in the training runtime template to apply the overrides.
 	// +listType=atomic
-	TargetJobs []PodSpecOverrideTargetJob `json:"targetJobs"`
+	TargetJobs []PodTemplateOverrideTargetJob `json:"targetJobs"`
 
+	// Override for the Pod template metadata.
+	// These values will be merged with the TrainingRuntime's Pod template metadata.
+	Metadata *PodTemplateMetadataOverride `json:"metadata,omitempty"`
```

Comment:
```suggestion
	metav1.ObjectMeta `json:"metadata,omitempty"`
```

It would be good to use metav1.ObjectMeta, which is a standardized way in Kubernetes.

Reply:
Isn't meant to to narrow the scope of what can be overwritten?

Reply:
I'd be inclined to be conservative here and grow when needs be. That'll also avoid potentially adding new fields implicitly when upgrading Kubernetes API.

Reply:
> If we allow overriding the entire spec, we need to ensure that certain .spec fields cannot be changed (e.g., the container name or the Trainer container resources).

I didn't indicate that having an entire `.spec`. The spec still has `PodSpecOverride` instead of an entire PodSpec.

```
podTemplateSpecOverride:
  targetJobs: # []PodTemplateSpecOverrideTargetJob
  template:
    metadata: # ObjectMeta
    spec: # PodSpecOverride w/o TargetJobs
```

> I'd be inclined to be conservative here and grow when needs be. That'll also avoid potentially adding new fields implicitly when upgrading Kubernetes API.

This implies that we may evolve the `metadata` field with arbitrary fields like `metadata.xyzTargetTimestamp`. However, I think avoiding custom metadata semantics would be better.

So, using `ObjectMeta` allows us to avoid such situations in the future, which could be considered as a restriction for API expansion. 

In the case of `PodTemplateMetadataOverride`, it opens for both programming specifications (how can we handle and which fields are propagated to which underlying objects (e.g., JobSet)) and API specifications.

In the case of `ObjectMeta`, it opens only for the programming specifications and the relaxiness of its programming specifications is restricted, rather than `PodTemplateMetadataOverride`, since the `ObjectMeta` is idiomatic typed and operations could be aligned with the existing ones.

Reply:
I synced with @tenzen-y and the proposal maps onto (2.) or (3.) from my list. Since (3.) has intrinsic gaps as k8s evolves the validation will never be perfect.

I think it might be surprising for users to see different approaches to metadata and spec. However as Yuki said it is for historical reasons. So, IIUC spec can be also adjusted in a follow up PR while the API remains in Alpha.

Reply:
@mimowo Adjusted to use built-in Metadata.